### PR TITLE
(GH-1038) Halt operation if there is a pending reboot already existing, or from package dependency

### LINKED ISSUE #1038: Feature/switch to halt install if dependency requires a reboot
As per https://groups.google.com/forum/#!topic/chocolatey/jP4cv8TDUcU, it's probably a good idea to add a switch so that install/upgrade will halt if a dependency requires a reboot. For example, I have a package that has a dependency on PowerShell 3+ - chocolately installed PowerShell v5 but then logged this and then continued to install packages with PowerShell 2 (and subsequently failed):

2016-11-02 12:05:55,829 [INFO ] - PowerShell.5.0 has been installed.
2016-11-02 12:05:55,829 [INFO ] -
2016-11-02 12:05:55,844 [WARN ] - WARNING: PowerShell.5.0 requires a reboot to complete the installation.
2016-11-02 12:05:55,844 [DEBUG] - Built-in PowerShell host called with ['[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\PowerShell\tools\ChocolateyInstall.ps1' -installArguments '' -packageParameters '''] exited with '3010'.
2016-11-02 12:05:55,844 [DEBUG] - Calling command ['"C:\Windows\System32\shutdown.exe" /a']
2016-11-02 12:05:55,870 [DEBUG] - Command ['"C:\Windows\System32\shutdown.exe" /a'] exited with '1116'

### PR DESCRIPTION
**Pull request by gep13**
Fixes #1038

### CODE REVIEW

**Comment by gep13** [comment]
This is an initial implementation of what is required, and this PR is being opened for review/comment.  The current implementation results in the following:

image

when attempting to install a package, where one of it's dependencies results in a reboot request.

File: src/chocolatey/infrastructure.app/builders/ConfigurationBuilder.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2017 - 2018 Chocolatey Software, Inc
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/builders/ConfigurationBuilder.cs]: UTF 8?

File: src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs [hunk 2]
```diff
@@ -155,6 +155,11 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
                      "Stop On First Package Failure - stop running install, upgrade or uninstall on first package failure instead of continuing with others. Overrides the default feature '{0}' set to '{1}'. Available in 0.10.4+.".format_with(ApplicationParameters.Features.StopOnFirstPackageFailure, configuration.Features.StopOnFirstPackageFailure.to_string()),
                      option => configuration.Features.StopOnFirstPackageFailure = option != null
                      )
+                 .Add("stoponfirstrebootrequest|stop-on-first-reboot-request|stop-on-first-package-reboot-request",
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: `exitwhenrebootdetected|exit-when-reboot-detected`

Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: This applies to Upgrade and uninstall as well.

File: src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs [hunk 3]
```diff
@@ -155,6 +155,11 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
                      "Stop On First Package Failure - stop running install, upgrade or uninstall on first package failure instead of continuing with others. Overrides the default feature '{0}' set to '{1}'. Available in 0.10.4+.".format_with(ApplicationParameters.Features.StopOnFirstPackageFailure, configuration.Features.StopOnFirstPackageFailure.to_string()),
                      option => configuration.Features.StopOnFirstPackageFailure = option != null
                      )
+                 .Add("stoponfirstrebootrequest|stop-on-first-reboot-request|stop-on-first-package-reboot-request",
+                     "Stop On First Package Reboot Request - stop running install, upgrade, or uninstall on first package reboot request instead of continuing with others.  Overrides the default feature '{0}' set to '{1}'.  Available in 0.10.12+.".format_with
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: "Exit When Reboot Detected -" and similar messaging that the feature switch will have.

File: src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs [hunk 4]
```diff
@@ -155,6 +155,11 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
                      "Stop On First Package Failure - stop running install, upgrade or uninstall on first package failure instead of continuing with others. Overrides the default feature '{0}' set to '{1}'. Available in 0.10.4+.".format_with(ApplicationParameters.Features.StopOnFirstPackageFailure, configuration.Features.StopOnFirstPackageFailure.to_string()),
                      option => configuration.Features.StopOnFirstPackageFailure = option != null
                      )
+                 .Add("stoponfirstrebootrequest|stop-on-first-reboot-request|stop-on-first-package-reboot-request",
+                     "Stop On First Package Reboot Request - stop running install, upgrade, or uninstall on first package reboot request instead of continuing with others.  Overrides the default feature '{0}' set to '{1}'.  Available in 0.10.12+.".format_with
+                     (ApplicationParameters.Features.StopOnFirstPackageRebootRequest, configuration.Features.StopOnFirstPackageRebootRequest.to_string()),
+                     option => configuration.Features.StopOnFirstPackageRebootRequest = option != null
+                     )
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: We'll also need `ignoredetectedreboot|ignore-detected-reboot`.

Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: Similar to https://github.com/chocolatey/choco/commit/6b446a5a5410d9b5061209832f6e39d6ac63cb41

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs [hunk 5]
```diff
@@ -420,6 +420,19 @@ public void handle_package_result(PackageResult packageResult, ChocolateyConfigu
 
             remove_pending(packageResult, config);
 
+            // Test to see if exit code is either:
+            // 1641 - restart initiated
+            // 3010 - restart required
+            var rebootExitCodes = new List<int> { 1641, 3010 };
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: 350 and 1604

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs [hunk 6]
```diff
@@ -420,6 +420,19 @@ public void handle_package_result(PackageResult packageResult, ChocolateyConfigu
...
+            // Test to see if exit code is either:
+            // 1641 - restart initiated
+            // 3010 - restart required
+            var rebootExitCodes = new List<int> { 1641, 3010 };
+
+            if(rebootExitCodes.Contains(packageResult.ExitCode))
+            {
+                if(config.Features.StopOnFirstPackageRebootRequest)
+                {
+                    throw new ApplicationException("Stopping further execution as {0} has requested a reboot.".format_with(packageResult.Name));
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: We need a way to stop completely if a package has requested for this. I'm wondering if NuGet Service might be a better place to exit the for loop - this is only for install/upgrade if I recall correctly. You will need to find the uninstall result one as well.

File: src/chocolatey.console/Program.cs [hunk 7]
```diff
@@ -158,7 +158,7 @@ private static void Main(string[] args)
                     "chocolatey".Log().Error(ChocolateyLoggers.Important, () => "{0}".format_with(ex.Message));
                 }
 
-                Environment.ExitCode = 1;
+                if (Environment.ExitCode == 0) Environment.ExitCode = 1;
```
Comment (ferventcoder) [inline src/chocolatey.console/Program.cs]: This is a good fix, glad we found this!

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs [hunk 8]
```diff
@@ -85,9 +88,63 @@ public void run(string[] args, ChocolateyConfiguration config, Container contain
                                 }
                             }
                         },
-                        () => command.handle_validation(config),
+                        () =>
+                        {
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Formatting, bump this up to the end of the first line to follow consistency with the rest of these

Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: ` () => {`

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs [hunk 9]
```diff
@@ -85,9 +88,63 @@ public void run(string[] args, ChocolateyConfiguration config, Container contain
                                 }
                             }
                         },
-                        () => command.handle_validation(config),
+                        () =>
+                        {
+                            command.handle_validation(config);
+
+                            var validationResults = new List<ValidationResult>();
+                            var validations = container.GetAllInstances<IValidation>();
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Maybe call this `validationChecks` or something, I'm not super keen on `validations`.

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs [hunk 10]
```diff
@@ -85,9 +88,63 @@ public void run(string[] args, ChocolateyConfiguration config, Container contain
...
+                            foreach (var validation in validations)
+                            {
+                                validationResults.AddRange(validation.validate(config));
+                            }
+
+                            var validationErrors = report_validation_summary(validationResults);
+
+                            if (validationErrors != 0)
+                            {
+                                // NOTE: This is intentionally left blank
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Also mention why this is left blank ("as we reported the summary above")

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs [hunk 11]
```diff
@@ -85,9 +88,63 @@ public void run(string[] args, ChocolateyConfiguration config, Container contain
...
+                }
+            }
+
+            if (errors != 0)
+            {
+                this.Log().Info("");
+                this.Log().Error("Errors:");
+                foreach (var error in validationResults.Where(p => p.Status == ValidationStatus.Error).or_empty_list_if_null())
+                {
+                    this.Log().Error(" - {0}".format_with(error.Message, error.ExitCode));
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Your format with has 2 items, but you are only passing a single item. Perhaps you meant to do this after copy pasta? 

`this.Log().Error(" - {0} ({1})".format_with(error.Message, error.ExitCode));`

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs [hunk 12]
```diff
@@ -85,9 +88,63 @@ public void run(string[] args, ChocolateyConfiguration config, Container contain
...
+                }
+            }
+
+            return errors;
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Overall good summary. However I think I might bump this up so the messaging looks like this:

~~~
5 validations performed. 2 success(es), 1 warning(s), and 2 error(s).
 Warnings:
   - Message here
 Errors:
   - Error message here (exit 350).
   - Next Error message here (exit 2).
~~~

Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: If the words "Warnings" and "Errors" do have a line break, they need to be prefixed with "Validation" as in "Validation Warnings"/"Validation Errors"

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs [hunk 13]
```diff
@@ -420,6 +420,24 @@ public void handle_package_result(PackageResult packageResult, ChocolateyConfigu
...
+
+            if(rebootExitCodes.Contains(packageResult.ExitCode))
+            {
+                if(config.Features.ExitOnRebootDetected)
+                {
+                    Environment.ExitCode = 1604;
+                    this.Log().Warn(ChocolateyLoggers.Important, @"Chocolatey has detected a pending reboot after installing/upgrading
+package '{0}' - stopping further execution".format_with(packageResult.Name));
+
+                    // NOTE: This is intentionally left blank
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: As above, add the reasoning here as well.

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs (hunk 10 above)
Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Was thinking on this after we set it that way and I think a reason would be helpful when we are in there later and see the note but then go to ask why.

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs [hunk 14]
```diff
@@ -560,29 +578,34 @@ private string capture_arguments(ChocolateyConfiguration config, PackageResult p
...
-            }
+                var installFailures = report_action_summary(packageInstalls, "installed");
+                if (installFailures != 0 && Environment.ExitCode == 0)
+                {
+                    Environment.ExitCode = 1;
+                }
 
-            randomly_notify_about_pro_business(config);
+                randomly_notify_about_pro_business(config);
+            }
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: This is a good start with install, but you'll also need this with upgrade and uninstall as well.

File: src/chocolatey/infrastructure.app/validations/GlobalConfigurationValidation.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+                    ExitCode = 1
+                };
+            }
+            else
+            {
+                return new ValidationResult
+                {
+                    Message = "Global Chocolatey Configuration is valid.",
+                    Status = ValidationStatus.Success,
+                    ExitCode = 0
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/validations/GlobalConfigurationValidation.cs]: This is incorrect to me. The validate method should see there are no warnings or issues and then return a status like this. I wouldn't do it here as once we add more, this won't work the way it is written.

File: src/chocolatey/infrastructure.app/validations/SystemStateValidation.cs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+                    };
+                }
+            }
+            else
+            {
+                return new ValidationResult
+                {
+                    Message = "System State is valid.",
+                    Status = ValidationStatus.Success,
+                    ExitCode = 0
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/validations/SystemStateValidation.cs]: Same idea here - loop over the results after the checks in the `validate()` method and add this if there are no warnings/errors.

File: src/chocolatey/infrastructure.app/validations/GlobalConfigurationValidation.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+                return new ValidationResult
+                {
+                    Message = @"When attempting to halt execution of a Chocolatey command based on a
+request for a system reboot, it is necessary to have the
+usePackageExitCodes feature enabled.  Currently, this is disabled. This
+can be re-enabled using the command:
+
+choco feature enable -name={0}
+".format_with(ApplicationParameters.Features.UsePackageExitCodes),
+                    Status = ValidationStatus.Error,
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/validations/GlobalConfigurationValidation.cs]: This should also check what we are attempting to run and warn on most commands - error on install/upgrade/uninstall.

File: src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs (hunk 11 above)
Comment (gep13) [reply src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Ah, good catch.  Original, I was outputting the exit code at this point as well, but then I noticed that this was already handled, and didn't need it anymore.

Comment (gep13) [reply src/chocolatey/infrastructure.app/runners/ConsoleApplication.cs]: Based on the comment below, I will add this back in though.

**Comment by gep13** [comment]
@ferventcoder I believe that this is ready for another review.  I have one question remaining about the flow of the exceptions after some more testing this morning.  But I will catch up with you regarding this later.

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs (hunk 14 above)
Comment (gep13) [reply src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: This was added into the uninstall/upgrade sections as well.

File: src/chocolatey/infrastructure.app/validations/SystemStateValidation.cs (hunk 16 above)
Comment (gep13) [reply src/chocolatey/infrastructure.app/validations/SystemStateValidation.cs]: In the end, I didn't enumerate, but rather simply count the results.  If there are 0, add the success message.

**Comment by gep13** [comment]
@ferventcoder updated again based on our conversations.  I believe I have caught everything that was discussed.

File: src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs [hunk 18]
```diff
@@ -155,6 +155,21 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
...
                      option => configuration.Features.StopOnFirstPackageFailure = option != null
                      )
+                 .Add("exitwhenrebootdetected|exit-when-reboot-detected",
+                     "Exit When Reboot Detected - Stop running install, upgrade, or uninstall when a reboot request is detected. Requires '{0}' feature to be turned on. Will exit with either {1} or {2}.  Overrides the default feature '{3}' set to '{4}'.  Available in 0.10.12+.".format_with
+                     (ApplicationParameters.Features.UsePackageExitCodes, ApplicationParameters.ExitCodes.ErrorFailNoActionReboot, ApplicationParameters.ExitCodes.ErrorInstallSuspend, ApplicationParameters.Features.ExitOnRebootDetected, configuration.Features.ExitOnRebootDetected.to_string()),
+                     option => configuration.Features.ExitOnRebootDetected = option != null
+                     )
+                 .Add("ignoredetectedreboot|ignore-detected-reboot",
+                     "Ignore Detected Reboot - If a reboot request is detected during a Chocolatey operation, then ignore it. Overrides the default feature '{0}' set to '{1}'.  Available in 0.10.12+.".format_with
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyInstallCommand.cs]: The comma throws me off. That is not a pause point. I would change the wording here "Ignore any detected reboots if found"

File: src/chocolatey/infrastructure.app/commands/ChocolateyUninstallCommand.cs [hunk 19]
```diff
@@ -118,6 +118,21 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
...
                      option => configuration.Features.StopOnFirstPackageFailure = option != null
                      )
+                 .Add("exitwhenrebootdetected|exit-when-reboot-detected",
+                     "Exit When Reboot Detected - Stop running install, upgrade, or uninstall when a reboot request is detected. Requires '{0}' feature to be turned on. Will exit with either {1} or {2}.  Overrides the default feature '{3}' set to '{4}'.  Available in 0.10.12+.".format_with
+                         (ApplicationParameters.Features.UsePackageExitCodes, ApplicationParameters.ExitCodes.ErrorFailNoActionReboot, ApplicationParameters.ExitCodes.ErrorInstallSuspend, ApplicationParameters.Features.ExitOnRebootDetected, configuration.Features.ExitOnRebootDetected.to_string()),
+                     option => configuration.Features.ExitOnRebootDetected = option != null
+                     )
+                 .Add("ignoredetectedreboot|ignore-detected-reboot",
+                     "Ignore Detected Reboot - If a reboot request is detected during a Chocolatey operation, then ignore it. Overrides the default feature '{0}' set to '{1}'.  Available in 0.10.12+.".format_with
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyUninstallCommand.cs]: Same comment as earlier

File: src/chocolatey/infrastructure.app/commands/ChocolateyUpgradeCommand.cs [hunk 20]
```diff
@@ -182,6 +182,21 @@ public virtual void configure_argument_parser(OptionSet optionSet, ChocolateyCon
...
                     })
+                 .Add("exitwhenrebootdetected|exit-when-reboot-detected",
+                     "Exit When Reboot Detected - Stop running install, upgrade, or uninstall when a reboot request is detected. Requires '{0}' feature to be turned on. Will exit with either {1} or {2}.  Overrides the default feature '{3}' set to '{4}'.  Available in 0.10.12+.".format_with
+                         (ApplicationParameters.Features.UsePackageExitCodes, ApplicationParameters.ExitCodes.ErrorFailNoActionReboot, ApplicationParameters.ExitCodes.ErrorInstallSuspend, ApplicationParameters.Features.ExitOnRebootDetected, configuration.Features.ExitOnRebootDetected.to_string()),
+                     option => configuration.Features.ExitOnRebootDetected = option != null
+                     )
+                 .Add("ignoredetectedreboot|ignore-detected-reboot",
+                     "Ignore Detected Reboot - If a reboot request is detected during a Chocolatey operation, then ignore it. Overrides the default feature '{0}' set to '{1}'.  Available in 0.10.12+.".format_with
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/commands/ChocolateyUpgradeCommand.cs]: same comment as earlier

File: src/chocolatey/infrastructure.app/registration/ContainerBinding.cs [hunk 21]
```diff
@@ -126,6 +129,19 @@ public void RegisterComponents(Container container)
...
+                {
+                    var list = new List<IValidation>
+                    {
+                        new GlobalConfigurationValidation(),
+                        new SystemStateValidation(container.GetInstance<IPendingRebootService>())
+                    };
+
+                    return list.AsReadOnly();
+                },
+                Lifestyle.Singleton);
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/registration/ContainerBinding.cs]: Validations Engine is an entirely different commit. I would separate the validation engine from the reboot detection.

I would also separate the commits for Pending Reboot detection versus Package Exit Code indicating Reboot.

Comment (ferventcoder) [reply src/chocolatey/infrastructure.app/registration/ContainerBinding.cs]: For global validations, let's even create an issue to tie those commits to. It's an interesting addition and we want to be able to tell folks when we first implemented global validations. Much harder to do if there is not an issue covering it.

File: src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs [hunk 22]
```diff
@@ -420,6 +420,23 @@ public void handle_package_result(PackageResult packageResult, ChocolateyConfigu
 
             remove_pending(packageResult, config);
 
+            // Test to see if exit code is either:
+            // 1641 - restart initiated
+            // 3010 - restart required
+            var rebootExitCodes = new List<int> { 1641, 3010 };
```
Comment (ferventcoder) [inline src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs]: Minor nitpick on memory/perf - we'd create this list every single time this method is called. Can we move this out to a class level private variable so it is only created once?

**Comment by gep13** [comment]
@ferventcoder I have raised a related issue here: https://github.com/chocolatey/choco/issues/1746

**Comment by gep13** [comment]
@ferventcoder I have broken the single commit into three separate ones, in a way that I think makes sense, but let me know if there is anything that you would like to change.

**Comment by ferventcoder** [comment]
Lost my merge commit on the rebase from stable. This was merged into stable at https://github.com/chocolatey/choco/commit/964d7d049170e6d408e7188db763ab2d2681bb3a and will be in 0.10.12. I added two additional commits for formatting the logging and xml comments.

**Pull request closed without merging**